File: backend/app/api/v1/endpoints/admin_assistance.py
"""Focused global-administrator assistance operations."""

from datetime import datetime
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.api.v1.endpoints.auth import get_current_admin_user
from app.db.database import get_db
from app.models.user import User
from app.services.raffle_assistance_service import RaffleAssistanceError, RaffleAssistanceService

router = APIRouter()


class RaffleRescheduleRequest(BaseModel):
    local_scheduled_at: datetime
    timezone_name: str = Field(min_length=1, max_length=64)
    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=5, max_length=1000)
    explicit_confirmation: bool
    snapshot_decision: Literal["preserve", "invalidate"] = "preserve"


def _error(exc: RaffleAssistanceError) -> HTTPException:
    status = 404 if exc.code == "raffle_not_found" else 409 if exc.code not in {"invalid_public_code", "invalid_timezone", "nonexistent_local_time", "ambiguous_local_time", "schedule_in_past"} else 422
    return HTTPException(status, detail={"code": exc.code, "message": exc.message})


@router.get("/raffles/lookup")
def lookup_raffle(identifier: str = Query(min_length=1, max_length=500), db: Session = Depends(get_db), _admin: User = Depends(get_current_admin_user)):
    try:
        return RaffleAssistanceService.serialize(db, RaffleAssistanceService.lookup(db, identifier))
    except RaffleAssistanceError as exc:
        raise _error(exc) from exc


@router.patch("/raffles/by-code/{public_code}/schedule")
def reschedule_raffle(public_code: str, payload: RaffleRescheduleRequest, db: Session = Depends(get_db), admin: User = Depends(get_current_admin_user)):
    try:
        raffle, audit = RaffleAssistanceService.reschedule(db, public_code=public_code, actor=admin, **payload.model_dump())
        db.commit(); db.refresh(raffle)
        return {"audit_id": audit.id, "raffle": RaffleAssistanceService.serialize(db, raffle)}
    except RaffleAssistanceError as exc:
        db.rollback(); raise _error(exc) from exc

File: frontend/src/services/adminAssistance.ts
import api from './api';

export interface AssistedRaffle {
  id: number;
  public_code: string;
  title: string;
  guild_name: string;
  purpose: string;
  status: string;
  execution_state: string;
  timezone_name: string;
  scheduled_run_at_utc: string | null;
  scheduled_run_at_local: string | null;
  participant_count: number;
  version: number;
  safe_to_reschedule: boolean;
  unsafe_reason: string | null;
  eligibility_snapshot: { exists: boolean; valid: boolean; warning: string | null; id: number | null };
  scheduler: { job_id: string | null; claimed_at: string | null; lease_expires_at: string | null; next_retry_at: string | null; retry_count: number; attempt_count: number; last_error_code: string | null };
}

export const adminAssistanceApi = {
  lookupRaffle: async (identifier: string): Promise<AssistedRaffle> =>
    (await api.get('/admin/assistance/raffles/lookup', { params: { identifier } })).data,
  rescheduleRaffle: async (publicCode: string, payload: {
    local_scheduled_at: string;
    timezone_name: string;
    expected_version: number;
    reason: string;
    explicit_confirmation: boolean;
    snapshot_decision: 'preserve' | 'invalidate';
  }): Promise<{ audit_id: number; raffle: AssistedRaffle }> =>
    (await api.patch(`/admin/assistance/raffles/by-code/${encodeURIComponent(publicCode)}/schedule`, payload)).data,
};
